src/demo/ContextMenuPopup.tsx

```diff
@@ -0,0 +1,113 @@
+ */
+export interface MenuItem {
+    id: string;
+    type: ContextMenuItemType;
+    caption: string;
+    hasChildren: boolean;
+}
+
+interface ContextMenuPopupProps {
+    items?: MenuItem[];
+    children?: React.ReactNode;
+}
+
+// tslint:disable-next-line:variable-name
+export const ContextMenuPopup: React.FunctionComponent<ContextMenuPopupProps> = (props) => {
+    const [menuItems, setMenuItems] = React.useState<MenuItem[]>([]);
+    let windowHeight = 2;
+
+    React.useEffect(() => {
+        const handleMenuMessage = (message: ContextMenuMessage) => {
+            if (message.action === "update") {
+                setMenuItems(message.items);
+                const ofWindow: fin.OpenFinWindow = fin.desktop.Window.getCurrent();
+                ofWindow.isShowing((showing) => {
+                    if (!showing) {
+                        ofWindow.show();
+                        ofWindow.focus();
+                    }
+                });
+            }
+        };
+        fin.desktop.InterApplicationBus.subscribe(uuid, ContextMenu.TOPIC, handleMenuMessage);
+
+        const listener = () => {
+            sendMessage({action: "blur"});
+        };
+        fin.desktop.Window.getCurrent().addEventListener("blurred", listener);
+
+        //Cleanup
+        return () => {
+            fin.desktop.InterApplicationBus.unsubscribe(uuid, ContextMenu.TOPIC, handleMenuMessage);
+            fin.desktop.Window.getCurrent().removeEventListener('blurred', listener);
+        };
+    }, []);
+
+    //Generate the menu items
+    const items = React.useMemo(() => {
+        return menuItems.map((item) => {
+            let component: JSX.Element;
+            let itemHeight = 22;
+            const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+                event.preventDefault();
+                onClick(item, event.currentTarget.offsetTop);
+            };
+
+            switch (item.type) {
+                case ContextMenuItemType.SPINNER:
+                    component = <button key={item.id} className="spinner w3-button w3-disabled"><i className="fa fa-spinner fa-spin" /></button>;
+                    break;
+                case ContextMenuItemType.SEPARATOR:
+                    component = <hr key={item.id} />;
+                    itemHeight = 7;
+                    break;
+                case ContextMenuItemType.LABEL:
+                    component = <span key={item.id} className="w3-button">{item.caption}</span>;
+                    break;
+                case ContextMenuItemType.BUTTON:
+                    component = <button key={item.id} className={"w3-button" + (item.hasChildren ? " children" : "")} onClick={handleClick}>{item.caption}</button>;
+                    break;
+                default:
+                    component = <></>;
+                    itemHeight = 0;
+            }
+
+            windowHeight += itemHeight;

```

Comment:
Can I check this is safe and/or 'The React Way'? Do we know for certain that this method will get called every time windowHeight is reset to 2 at the start of the component function?

Reply:
You're right, it doesn't get reset to 2. Fixed it.